### PR TITLE
feat(weave): backend for new costs query

### PR DESCRIPTION
**Pull request by jwlee64**
ahead of this I merged
orm updates to add join and groupby
moving the clickhouse schema objs to new file
add the llm_token_prices table

creates some query funcs in token_costs.py to add costs under summary.weave.costs to the calls query
enables costs in on the req include_costs = True

### CODE REVIEW

File: weave/trace_server/trace_server_interface.py [hunk 1]
```diff
@@ -206,6 +240,7 @@ class CallEndRes(BaseModel):
 class CallReadReq(BaseModel):
     project_id: str
     id: str
+    add_cost: typing.Optional[bool] = False
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:243]: nit: can we call this `include_costs`

File: weave/trace_server/trace_server_interface.py [hunk 2]
```diff
@@ -253,6 +288,7 @@ class CallsQueryReq(BaseModel):
     # Sort by multiple fields
     sort_by: typing.Optional[typing.List[_SortBy]] = None
     query: typing.Optional[Query] = None
+    add_cost: typing.Optional[bool] = False
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:291]: can you add a unit test that exercises this path - going through the client?

File: weave/trace_server/trace_server_interface.py [hunk 3]
```diff
@@ -19,11 +19,25 @@ class ExtraKeysTypedDict(TypedDict):
 # https://docs.pydantic.dev/2.8/concepts/strict_mode/#dataclasses-and-typeddict
 ExtraKeysTypedDict.__pydantic_config__ = ConfigDict(extra="allow")  # type: ignore
 
-
-class WeaveSummarySchema(ExtraKeysTypedDict):
-    status: typing.Optional[typing.Literal["success", "error", "running"]]
-    nice_trace_name: typing.Optional[str]
-    latency: typing.Optional[int]
+cost_string_fields = [
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:22]: I would recommend moving these lists into your DB layer. I think they are more descriptive of the on-disk state than the over-the-wire type

File: weave/trace_server/sqlite_trace_server.py [hunk 4]
```diff
@@ -1097,11 +1097,16 @@ def _make_derived_summary_map(
     status = make_call_status_from_exception_and_ended_at(exception, ended_at_dt)
     latency = None if not ended_at_dt else (ended_at_dt - started_at_dt).microseconds
     display_name = display_name or op_name_simple_from_ref_str(op_name)
+    summary = summary_dump or {}
```
Comment (tssweeney) [inline weave/trace_server/sqlite_trace_server.py:1100]: this is an ugly artifact of these typing shanagans - i think we should punt on that instead of adding all these nulls over the wire

File: weave/trace_server/calls_query_builder.py [hunk 5]
```diff
@@ -216,6 +217,7 @@ class CallsQuery(BaseModel):
     order_fields: list[OrderField] = Field(default_factory=list)
     limit: typing.Optional[int] = None
     offset: typing.Optional[int] = None
+    should_add_cost: bool = False
```
Comment (tssweeney) [inline weave/trace_server/calls_query_builder.py:220]: nit: `include_costs`

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 6]
```diff
@@ -352,13 +353,18 @@ def calls_query_stream(
...
+        cq.add_cost(req.add_cost or False)
 
         # TODO (Perf): By allowing a sub-selection of columns
         # we will gain increased performance by not having to
         # fetch all columns from the database. Currently all use
         # cases call for every column to be fetched, so we have not
         # implemented this yet.
-        columns = all_call_select_columns
+        # We put summary_dump last so that when we compute the costs and summary its in the right place
+        columns = [
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:364]: might want to be 
```
columns = all_call_select_columns
if req.add_cost:
   # move to end

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 7]
```diff
@@ -1425,12 +1431,17 @@ def _make_derived_summary_map(
         None if not ended_at or not started_at else (ended_at - started_at).microseconds
     )
     display_name = display_name or op_name_simple_from_ref_str(op_name)
+    summary = summary_dump or {}
+    if "_weave" not in summary:
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:1435]: i think this goes away if we drop the pydantic stuff

File: weave/trace_server/migrations/005_add_cost.up.sql [hunk 8]
```diff
@@ -0,0 +1,49 @@
+CREATE TABLE llm_token_prices (
```
Comment (tssweeney) [inline weave/trace_server/migrations/005_add_cost.up.sql:1]: how does this get populated? Should we have a mechanism of bootstrapping this / updating from some public location?

File: weave/trace_server/token_costs.py [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+    Column(name="completion_token_cost", type="float"),
+    Column(name="prompt_token_cost_unit", type="string"),
+    Column(name="completion_token_cost_unit", type="string"),
+]
+LLM_TOKEN_PRICES_TABLE_NAME = "llm_token_prices"
+LLM_TOKEN_PRICES_TABLE = Table(
+    name=LLM_TOKEN_PRICES_TABLE_NAME, cols=LLM_TOKEN_PRICES_COLUMNS
+)
+
+CALLS_MERGED_COLUMNS = [
```
Comment (tssweeney) [inline weave/trace_server/token_costs.py:20]: can we derive this from the other list to have a single source of truth?

File: weave/trace_server/migrations/005_add_cost.up.sql [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    /*
+    `completion_token_cost`: The cost of a completion token in the specified LLM.
+    */
+    completion_token_cost Float,
+
+    /*
+    `completion_token_cost_unit`: The unit of completion token cost in the specified LLM.
+    */
+    completion_token_cost_unit String
```
Comment (jamie-rasmussen) [inline weave/trace_server/migrations/005_add_cost.up.sql:46]: Do we need some fields for auditing purposes - which user inserted this row, when it was inserted?

File: weave/trace_server/clickhouse_schema.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
+import datetime
```
Comment (jwlee64) [inline weave/trace_server/clickhouse_schema.py:1]: moved from clickhouse_trace_server_batched. You can import there and token_costs, to keep ch definitions as 1 source of truth

File: weave/trace_server/trace_server_interface.py (hunk 2 above)
Comment (jwlee64) [reply weave/trace_server/trace_server_interface.py]: when I go through the client it hits the sqlite server rather than clickhouse, Is there something that I need to enable to correctly test through client?

Comment (tssweeney) [reply weave/trace_server/trace_server_interface.py]: every test that uses the `client` fixture will be run twice in CI - once for SQLite and once for CH. To force CH, you can run pytest with `--weave_server=clickhouse`. Checkout the `client` fixture for details

File: weave/weave_client.py [hunk 12]
```diff
@@ -310,7 +317,7 @@ def make_client_call(
         id=server_call.id,
         inputs=from_json(server_call.inputs, server_call.project_id, server),
         output=output,
-        summary=server_call.summary,
+        summary=dict(server_call.summary if server_call.summary else {}),
```
Comment (tssweeney) [inline weave/weave_client.py:320]: can you explain why this change is needed? Seems to effect some tests and I am curious what was the need for the here (perhaps left over from pydantic?)

File: weave/trace_server/trace_server_interface.py [hunk 13]
```diff
@@ -11,13 +12,67 @@
...
+
+class SummaryInsertMap(ExtraKeysTypedDict, total=False):
+    usage: typing.Dict[str, LLMUsageSchema]
+
+
+class SummaryMap(SummaryInsertMap, total=False):
+    weave: typing.Optional[WeaveSummarySchema]
+
+
+class WeaveAttributeSchema(TypedDict):
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:60]: unused?

File: weave/trace_server/trace_server_interface.py [hunk 14]
```diff
@@ -43,14 +98,20 @@ class CallSchema(BaseModel):
     output: typing.Optional[typing.Any] = None
 
     ## Summary: a summary of the call
-    summary: typing.Optional[typing.Dict[str, typing.Any]] = None
+    summary: typing.Optional[SummaryMap] = None
```
Comment (tssweeney) [inline weave/trace_server/trace_server_interface.py:101]: It seems like you found the golden solution here. Worth checking with @gtarpenning and @andrewtruong to check to see if the degenerate effects they noticed are exhibited here. In Griffin's PR we decided to punt on the pydantic typing for now. IDK how much trouble vs benefit this is causing you

File: weave/trace_server/migrations/005_add_cost.up.sql [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+    /*
+    `inserted_by`: User ID of the user who inserted the record, or the system if the record was inserted by the system.
+    */
+    inserted_by String,
+
+    /*
+    `inserted_at`: When the record was inserted.
+    */
+    inserted_at DateTime64(3) DEFAULT now64(3),
```
Comment (tssweeney) [inline weave/trace_server/migrations/005_add_cost.up.sql:56]: we use `created_at` in all other tables - can we use the same name here?

File: weave/trace_server/migrations/005_add_cost.up.sql [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+) ENGINE = ReplacingMergeTree()
```
Comment (tssweeney) [inline weave/trace_server/migrations/005_add_cost.up.sql:58]: why `ReplaceingMergeTree` over `MergeTree`?

File: weave/trace_server/update_cost_table.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
+# script to load data into llm_token_prices table in ClickHouse
```
Comment (tssweeney) [inline weave/trace_server/update_cost_table.py:1]: I think this routine might need some modifications to be production ready - perhaps it should be excluded from this PR?

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 18]
```diff
@@ -1496,7 +1405,7 @@ def _end_call_for_insert_to_ch_insertable_end_call(
         id=end_call.id,
         exception=end_call.exception,
         ended_at=end_call.ended_at,
-        summary_dump=_dict_value_to_dump(end_call.summary),
+        summary_dump=_dict_value_to_dump(dict(end_call.summary)),
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:1408]: curious what happens with the nested typed dicts? is `dict` enough to fix this? Should this be implemented in `_dict_value_to_dump`?

File: weave/trace_server/clickhouse_schema.py (hunk 11 above)
Comment (tssweeney) [reply weave/trace_server/clickhouse_schema.py:1]: nice - good idea. There will likely be some merge conflicts soon here, just fyi

File: weave/tests/test_weave_client.py [hunk 19]
```diff
@@ -305,8 +305,9 @@ def test_calls_query(client):
                 "os_version": platform.version(),
                 "os_release": platform.release(),
                 "sys_version": sys.version,
-            },
+            }
         },
+        summary={},
```
Comment (tssweeney) [inline weave/tests/test_weave_client.py:310]: seems odd to me that we now have an empty dict where we didn't before - is this intentional?

File: weave/trace_server/clickhouse_trace_server_batched.py (hunk 18 above)
Comment (jwlee64) [reply weave/trace_server/clickhouse_trace_server_batched.py:1408]: im going to move this to _any_value_to_dump and I confirmed this works for at least 5 deep nested dicts

File: weave/weave_client.py (hunk 12 above)
Comment (jwlee64) [reply weave/weave_client.py:320]: I can modify this ternary so we dont return {} and mess up tests, but this is necessary because we return SummaryMap, but i left call.summary as a dict, because the rollup of arbritrary keys got a bit complex using typeddicts

File: .github/workflows/test.yaml [hunk 20]
```diff
@@ -133,7 +133,7 @@ jobs:
           DD_SERVICE: weave-python
           DD_ENV: ci
           WEAVE_SENTRY_ENV: ci
-        run: CI=1 WB_SERVER_HOST=http://wandbservice WF_CLICKHOUSE_HOST=weave_clickhouse WEAVE_SERVER_DISABLE_ECOSYSTEM=1 source /root/venv/bin/activate && pip install 'dspy-ai>=0.1.5' && pip install pydantic -U  && cd weave && pytest --job-num=${{ matrix.job_num }} --timeout=90  ./flow ./integrations ./tests ./legacy/ops_arrow ./legacy/ecosystem ./trace_server ./trace --ddtrace --durations=5
+        run: CI=1 WB_SERVER_HOST=http://wandbservice WF_CLICKHOUSE_HOST=weave_clickhouse WEAVE_SERVER_DISABLE_ECOSYSTEM=1 source /root/venv/bin/activate && pip install 'dspy-ai>=0.1.5' && pip install pydantic -U  && cd weave && pytest -m "not skip_sqlite_client" --job-num=${{ matrix.job_num }} --timeout=90  ./flow ./integrations ./tests ./legacy/ops_arrow ./legacy/ecosystem ./trace_server ./trace --ddtrace --durations=5
```
Comment (jwlee64) [inline .github/workflows/test.yaml:136]: follows the pattern of `skip_clickhouse_client` adding costs only works in clickhouse as of now, because of all the json and array functions I do to calculated the costs. 

I probably could add the json stuff to the orm and make it work for both sqlite and clickhouse, but that is not an insignificant lift. For now I thought it better to write two tests, one for sqlite to make sure include_costs does nothing, and another for clickhouse that include_costs adds a cost object

File: weave/tests/test_weave_client.py [hunk 21]
```diff
@@ -1165,3 +1166,96 @@ def test_table_partitioning(network_proxy_client):
...
+    callsNoCost = list(client.calls(include_costs=False))
+
+    assert len(callsWithCost) == len(callsNoCost)
+    assert len(callsWithCost) == 4
+
+    noCostCallSummary = callsNoCost[0].summary
+    withCostCallSummary = callsWithCost[0].summary
+
+    assert withCostCallSummary.get("weave", "bah") != "bah"
+    assert len(withCostCallSummary["weave"]["costs"]) > 0
```
Comment (tssweeney) [inline weave/tests/test_weave_client.py:1231]: is this the only place where you are validating that your cost calculation is correct? Are there any other places where maybe you test:
* mixed models
* models that have 2 entries in the cost table over time,
* ...

The primary end-results?

File: weave/trace_server/clickhouse_trace_server_batched.py [hunk 22]
```diff
@@ -1397,7 +1406,7 @@ def _end_call_for_insert_to_ch_insertable_end_call(
         id=end_call.id,
         exception=end_call.exception,
         ended_at=end_call.ended_at,
-        summary_dump=_dict_value_to_dump(end_call.summary),
+        summary_dump=_any_value_to_dump(end_call.summary),
```
Comment (tssweeney) [inline weave/trace_server/clickhouse_trace_server_batched.py:1409]: this is not quite right. We want to ensure that `summary_dump` is in fact a dictionary dump. so `_dict_value_to_dump` is correct. I am not sure what is the motivation for this change as it opens us up to potentially non-dict data

Comment (tssweeney) [reply weave/trace_server/clickhouse_trace_server_batched.py:1409]: i think you want what was here before: `_dict_value_to_dump(dict(end_call.summary))`. 

I think the key here is that end_call.summary is now Expected to be a TypedDict after your typing changes.

File: weave/trace_server/migrations/006_seed_costs.up.sql [hunk 23]
```diff
@@ -0,0 +1,506 @@
+INSERT INTO llm_token_prices 
+    (id, pricing_level, pricing_level_id, provider_id, llm_id, effective_date, prompt_token_cost, prompt_token_cost_unit, completion_token_cost, completion_token_cost_unit, created_by, created_at) 
+VALUES 
+    (generateUUIDv4(), 'default', 'default', 'openai', 'gpt-4', now(), 3e-05, 'USD', 6e-05, 'USD', 'system', now()),
```
Comment (tssweeney) [inline weave/trace_server/migrations/006_seed_costs.up.sql:4]: please use UUIDv7 https://clickhouse.com/docs/en/sql-reference/functions/uuid-functions#generateUUIDv7

File: weave/trace_server/migrations/005_add_cost.up.sql [hunk 24]
```diff
@@ -1,4 +1,8 @@
 CREATE TABLE llm_token_prices (
+     /*
+    `id`: The unique identifier for the call. This is typically a UUID.
+    */
+    id String,
```
Comment (tssweeney) [inline weave/trace_server/migrations/005_add_cost.up.sql:5]: just flagging - very bad to add changes to a prior migration - i know we just landed this, but better form is to alter the table in yet another migration. Else, any commits between the land of the first table and now are invalid to deploy

File: weave/tests/test_weave_client.py [hunk 25]
```diff
@@ -1165,3 +1166,167 @@ def test_table_partitioning(network_proxy_client):
...
+    noCostCallSummary = callsNoCost[0].summary
+    withCostCallSummary = callsWithCost[0].summary
+
+    assert withCostCallSummary.get("weave", "bah") != "bah"
+    assert len(withCostCallSummary["weave"]["costs"]) == 2
+
+    gpt4cost = withCostCallSummary["weave"]["costs"]["gpt-4"]
+    gpt4ocost = withCostCallSummary["weave"]["costs"]["gpt-4o"]
+
+    # rounds to nearest integer avoids .999999992345234
```
Comment (tssweeney) [inline weave/tests/test_weave_client.py:1254]: tip: use `pytest.approx` - it is used in various parts of the code base for reference.

File: weave/trace_server/migrations/006_seed_costs.up.sql (hunk 23 above)
Comment (jwlee64) [reply weave/trace_server/migrations/006_seed_costs.up.sql:4]: i believe our version of clickhouse is too old for v7

Comment (tssweeney) [reply weave/trace_server/migrations/006_seed_costs.up.sql:4]: whaaa? according to what?

**Review by tssweeney — APPROVED** [review]
